cluster-autoscaler/cloudprovider/azure/azure_cache.go

```diff
@@ -353,6 +378,46 @@ func (m *azureCache) getAutoscalingOptions(ref azureRef) map[string]string {
 	return m.autoscalingOptions[ref]
 }
 
+// HasInstance returns if a given instance exists in the azure cache
+func (m *azureCache) HasInstance(providerID string) (bool, error) {
+	vmsPoolSet := m.getVMsPoolSet()
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	resourceID, err := convertResourceGroupNameToLower(providerID)
+	if err != nil {
+		// Most likely an invalid resource id, we should return false
+		// most of these shouldn't make it here do to higher level
+		// validation in the HasInstance azure.cloudprovider function
+		return false, err
+	}
+
+	nodeGroupNameForNode := getVMSSNameFromProviderID(providerID)
+	inst := azureRef{Name: resourceID}
+	klog.V(4).Infof("HasInstance: Checking instance %s", inst.Name)
+	if m.unownedInstances[inst] || (m.vmType == vmTypeVMSS && len(vmsPoolSet) == 0 && !m.registeredAKSPoolNames.Has(nodeGroupNameForNode)) {
+		klog.V(4).Infof("HasInstance: Instance %s is unowned", inst.Name)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if unowned, err := m.isUnownedInstance(inst, ""); unowned || err != nil {
+		klog.V(4).Infof("HasInstance: Instance %s is unowned or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if invalid, err := m.isInvalidVMType(inst, m.vmType, vmsPoolSet); invalid || err != nil {
+		klog.V(4).Infof("HasInstance: Instance %s is invalid VM type or error occurred: %v", inst.Name, err)
+		return false, cloudprovider.ErrNotImplemented
+	}
+
+	if nodeGroup := m.getInstanceFromCache(inst.Name); nodeGroup != nil {
+		klog.V(4).Infof("FindForInstance: found node group %q in cache", nodeGroup.Id())
+		return true, nil
```

Comment:
Why don't we use `return FindForInstance(...) != nil`?

Reply:
There is some different logic for checking for unowned instances.